[https://nvbugs/5464088] [fix] Guard against fp8 activations in lora forward; update perf test config

venkywonka opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Automatically derives LoRA count from model labels and sets max_loras and max_cpu_loras accordingly.
- Bug Fixes
  - Prevents runtime failures when LoRA receives unsupported input types by auto-casting to FP16/BF16, with a debug log for visibility.
- Tests
  - Perf configuration updated to reflect dynamic LoRA count and new presets, improving flexibility for PyTorch model runs.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up
-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

venkywonka committed:
[5464088][fix] Enhance LoRA support in PyTorch model configuration

- Added logging for dtype casting in LoraLayer to ensure compatibility with FP16/BF16.
- Updated model configuration to derive the number of LoRA adapters from the model label, improving flexibility in adapter management.

Signed-off-by: Venky Ganesh <23023424+venkywonka@users.noreply.github.com>

venkywonka commented:
/bot run

tensorrt-cicd commented:
[PR_Github #15700](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15700/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15700](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15700/) [ run ] completed with state `FAILURE`
[/LLM/release-1.0/L0_MergeRequest_PR pipeline #200](https://nv/trt-llm-cicd/job/release-1.0/job/L0_MergeRequest_PR/200/) completed with status: 'FAILURE'

venkywonka commented:
/bot run

tensorrt-cicd commented:
[PR_Github #15719](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15719/) [ run ] triggered by Bot

amitz-nv commented on tensorrt_llm/_torch/peft/lora/layer.py:
```diff
@@ -119,6 +121,15 @@ def forward(
             else:
+                # Guard: LoRA custom op only supports FP16/BF16 activations.
+                # If upstream produced FP8 (e.g., FP8 SwiGLU), cast here to avoid runtime failure.
+                if x.dtype not in (torch.float16, torch.bfloat16):
```
Is there a test that verifies this fixes the issue?

JunyiXu-nv replied:
This is kind of werid. The input type of the down_proj will be checked before calling the kernel, as we can see here: 

https://github.com/NVIDIA/TensorRT-LLM/blob/655d0f48d0d2d0659c54e86db87a56a3a0229f1e/tensorrt_llm/_torch/modules/gated_mlp.py#L103

Maybe we should revise the logic here to avoid using fp8 activation? I was using the `self.down_proj.has_fp8_qdq` to determine whether we use fp8 activation. Does the model set this argument here?

The input type will keep the same as the input `x` if no dtype is set when calling swiglu.

https://github.com/NVIDIA/TensorRT-LLM/blob/655d0f48d0d2d0659c54e86db87a56a3a0229f1e/tensorrt_llm/_torch/modules/swiglu.py#L62

venkywonka replied:
> I was using the self.down_proj.has_fp8_qdq to determine whether we use fp8 activation

@JunyiXu-nv , yes `self.down_proj.has_fp8_qdq` evaluates to true, but it shouldn't, when using lora (because our bindings dont' support it, not because it can't be quantized).

https://github.com/NVIDIA/TensorRT-LLM/blob/655d0f48d0d2d0659c54e86db87a56a3a0229f1e/tensorrt_llm/_torch/modules/swiglu.py#L62

oh, didn't notice that. then that would be a simpler place to modify is it? if we just override the `dtype` there in swigly, to ensure the outputs are half-precision instead of fp8 (specifically in lora path)?

venkywonka replied:
@JunyiXu-nv , so would you say its better to make this modification/change at the site of `swiglu` : 
```python
# in gated_mlp.py
def _apply_activation(self, x, *, for_lora: bool = False):
    if self.activation == F.silu:
        if self.down_proj.has_fp8_qdq:
            if for_lora:
                target = torch.bfloat16 if torch.cuda.is_bf16_supported() else torch.float16
                return swiglu(x, quant_scale=self.down_proj.input_scale, quant_type=target)
            else:
                return swiglu(x, quant_scale=self.down_proj.input_scale, quant_type=torch.float8_e4m3fn)
        else:
            return swiglu(x)
    elif callable(self.activation):
        return self.activation(x)
    elif self.activation is None:
        return x
    else:
        raise NotImplementedError(...)
```

venkywonka replied:
@JunyiXu-nv in the latest commit i have done the same as you suggested. can you confirm if this is what you meant. confirmed that this fixes the bug too.

JunyiXu-nv replied:
Yes, that's what I mean. The performance could be better if we don't involve the redundent quant/dequant.

amitz-nv replied:
I've conducted benchmarks recently (even today) on `Llama-3.1-8B-Instruct-FP8` with `Llama-3-Chinese-8B-Instruct-v2-LoRA` adapter and it works, what's different in the test perf that it fails there?

venkywonka commented on tensorrt_llm/_torch/peft/lora/layer.py:
```diff
@@ -119,6 +121,15 @@ def forward(
+                # Guard: LoRA custom op only supports FP16/BF16 activations.
+                # If upstream produced FP8 (e.g., FP8 SwiGLU), cast here to avoid runtime failure.
+                if x.dtype not in (torch.float16, torch.bfloat16):
+                    target_dtype = torch.bfloat16 if torch.cuda.is_bf16_supported(
```
I'm not sure myself, but the tests usually have dummy loras which aren't low precision. 
But either way the `lora_grouped_gemm` [[these lines](https://github.com/NVIDIA/TensorRT-LLM/blob/82276167e6a67cae84a7904a211de82f1593b7b6/cpp/tensorrt_llm/thop/loraOp.cpp#L147-L151
)] only supports fp16 anyways.

amitz-nv replied:
TBH I don't think I know... maybe Shahar or Nave know

venkywonka replied:
for now, either way, the only dtype for activations that can _not break_ `lora_grouped_gemm` is `bf16/fp16` ... so we're limited by that issue. 
I have added simplified the guarding and added a NOTE + TODO for this.

resolving this hence

venkywonka committed:
Move activation handling to GatedMLP for LoRA compatibility

- Modified _apply_activation method to accept a for_lora flag, allowing for specific handling of activation during LoRA operations.
- Updated the call to _apply_activation in GatedMLP to pass the for_lora argument, ensuring correct behavior in LoRA scenarios.
- Removed unnecessary dtype casting checks in LoraLayer, simplifying the code.

Signed-off-by: Venky Ganesh <23023424+venkywonka@users.noreply.github.com>

venkywonka commented:
/bot run

venkywonka commented:
/bot run

tensorrt-cicd commented:
[PR_Github #15833](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15833/) [ run ] triggered by Bot

venkywonka commented:
/bot run

tensorrt-cicd commented:
[PR_Github #15851](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15851/) [ run ] triggered by Bot

shaharmor98 commented on tensorrt_llm/_torch/modules/gated_mlp.py:
```diff
@@ -95,12 +96,23 @@ def __init__(self,
-                              quant_type=torch.float8_e4m3fn)
+                if for_lora:
+
+                    target = torch.bfloat16 if torch.cuda.is_bf16_supported(
```
I think it can be really error-prone. If one expects to use FP8 and you cast it to BF16/FP16, it will most likely affect the final output, no?

venkywonka replied:
but we don't have have a choice right? as fp8 anyways will break in `lora_grouped_gemm()` inside `tensorrt_llm/_torch/peft/lora/layer.py`

venkywonka replied:
ideally, we should find a way to add `fp8` support to `lora_grouped_gemm()` and then remove this guard. i can add a TODO there, but tha's beyond the scope of `release/1.0`.

venkywonka replied:
also, before #6540 , the `swiglu()` was emitting fp16 activations only all this while, so the change is merely restoring old behavior (only for lora path) and *not* adding new behavior.

shaharmor98 commented on tensorrt_llm/_torch/modules/gated_mlp.py:
```diff
@@ -95,12 +96,23 @@ def __init__(self,
+
+                    target = torch.bfloat16 if torch.cuda.is_bf16_supported(
+                    ) else torch.float16
+                    logger.debug(
```
I'm not sure that adding a debug log level is sufficient here. TRTLLM has many debug logs, and I don't really expect anyone to catch this kind of log. It might also have serious implications in terms of the model's output. I can think of many correctness tests and use cases that could be negatively affected by this change, and it could be extremely hard for practically anyone to catch this forced casting.

shaharmor98 commented on tensorrt_llm/_torch/modules/gated_mlp.py:
```diff
@@ -95,12 +96,23 @@ def __init__(self,
             [2 * self.intermediate_size // mapping.tp_size])
 
-    def _apply_activation(self, x):
+    def _apply_activation(self, x, *, for_lora: bool = False):
```
nit, a matter of personal test, feel free to resolve- instead of `for_lora` --> `has_lora` ?

JunyiXu-nv commented on tensorrt_llm/_torch/modules/gated_mlp.py:
```diff
@@ -95,12 +96,23 @@ def __init__(self,
+                    logger.debug(
+                        f"GatedMLP._apply_activation: LoRA path active; forcing non-FP8 activation dtype {target} (keeping activations in bf16/fp16), layer_idx={self.layer_idx}"
+                    )
+                    return swiglu(x,
+                                  quant_scale=self.down_proj.input_scale,
+                                  quant_type=target)
```
Just realize that the quant_scale/quant_type should not be set here. The output from `gate_up_proj` should already be bf16/fp16, and set quant_scale will let the kernel multiply the scale with x, which would involve accuracy bug.

Maybe revise like this:

```suggestion
                    return swiglu(x)
```

Or change the above condition `if self.down_proj.has_fp8_qdq:` to `if self.down_proj.has_fp8_qdq and not for_lora:`

venkywonka replied:
thank you @JunyiXu-nv  for catching that! i'm not familiar with this area of the code, so missed it. This should ideally have been caught right?
I guess L0 didnt' catch it but do you know if there is any accuracy coverage in L1+ for such an accuracy bug? 

maybe i should run a post merge to make sure this PR doesn't break anything (even after the above suggestion)?

venkywonka commented on tests/integration/defs/perf/pytorch_model_config.py:
```diff
@@ -181,10 +181,19 @@ def get_model_yaml_config(model_label: str,
+        # Derive the requested number of adapters from model_label (segment like "loras:X")
+        lora_count = 1
+        for part in model_label.split('-'):
+            if part.startswith('loras:'):
```
the test and QA side guarantees this. that's the string format agreed upon in the way we dynamically parse the test perf test cases, see `tests/integration/defs/perf/test_perf.py::PerfTestConfig.to_string`

venkywonka commented on tests/integration/defs/perf/pytorch_model_config.py:
```diff
@@ -181,10 +181,19 @@ def get_model_yaml_config(model_label: str,
                 'lora_dir': lora_dirs if lora_dirs is not None else [],
-                'max_lora_rank': 64
+                'max_lora_rank': 64,
+                'max_loras': lora_count,
```
My understainding is the below behavior was after @amitz-nv 's #6220 
- Default (no `max_loras` specified): LoRA GPU cache is sized by `PeftCacheConfig.device_cache_percent` (default 0.02 = 2% of free GPU memory after engine load), with paging determined by `optimal_adapter_size/max_adapter_size`
- With `max_loras` specified: it overrides the percentage and sets `num_device_module_layer = max_lora_rank × num_lora_modules × max_loras`, sizing the cache to hold that many adapters.

so TLDR; 
if we don't override with `max_loras`, the gpu cache set by default by `device_cache_percent` is not enough when running `test_perf[llama_v3.1_8b_instruct_fp8-bench-pytorch-float8-input_output_len:128,128-loras:1-reqs:100-con:2-gpus:1` and so overriding makes it pass.

I'd say its a WAR and the underlying issue is we probably lack the recovery and fallback.

@amitz-nv how do you think we should track the above identified issue? i can create a bug for this if you think it makes sense?

venkywonka replied:
yep, its right below this comment :)

venkywonka committed:
apply review comments

Signed-off-by: Venky Ganesh <23023424+venkywonka@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #15948](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15948/) [ run ] completed with state `SUCCESS`
[/LLM/release-1.0/L0_MergeRequest_PR pipeline #241](https://nv/trt-llm-cicd/job/release-1.0/job/L0_MergeRequest_PR/241/) completed with status: 'FAILURE'

venkywonka commented:
/bot run --extra-stage "A100X-PyTorch-1"

tensorrt-cicd commented:
[PR_Github #15959](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15959/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15959](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15959/) [ run ] completed with state `SUCCESS`
[/LLM/release-1.0/L0_MergeRequest_PR pipeline #243](https://nv/trt-llm-cicd/job/release-1.0/job/L0_MergeRequest_PR/243/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #15979](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15979/) [ run ] triggered by Bot

shaharmor98 reviewed (approved):
LGTM
